core/embed/rust/src/ui/model_tt/component/button.rs

```diff
@@ -340,29 +340,114 @@ pub struct ButtonStyle {
 }
 
 impl<T> Button<T> {
-    pub fn left_right<F0, F1, R>(
+    pub fn cancel_confirm(
         left: Button<T>,
-        left_map: F0,
         right: Button<T>,
-        right_map: F1,
-    ) -> (Map<GridPlaced<Self>, F0>, Map<GridPlaced<Self>, F1>)
+        right_size_factor: usize,
+    ) -> CancelConfirm<
+        T,
+        impl Fn(ButtonMsg) -> Option<CancelConfirmMsg>,
+        impl Fn(ButtonMsg) -> Option<CancelConfirmMsg>,
+    >
     where
-        F0: Fn(ButtonMsg) -> Option<R>,
-        F1: Fn(ButtonMsg) -> Option<R>,
         T: AsRef<str>,
     {
-        const BUTTON_SPACING: i32 = 6;
+        let columns = 1 + right_size_factor;
         (
             GridPlaced::new(left)
-                .with_grid(1, 3)
-                .with_spacing(BUTTON_SPACING)
+                .with_grid(1, columns)
+                .with_spacing(theme::BUTTON_SPACING)
                 .with_row_col(0, 0)
-                .map(left_map),
+                .map(|msg| {
+                    (matches!(msg, ButtonMsg::Clicked)).then(|| CancelConfirmMsg::Cancelled)
+                }),
             GridPlaced::new(right)
-                .with_grid(1, 3)
-                .with_spacing(BUTTON_SPACING)
-                .with_from_to((0, 1), (0, 2))
-                .map(right_map),
+                .with_grid(1, columns)
+                .with_spacing(theme::BUTTON_SPACING)
+                .with_from_to((0, 1), (0, right_size_factor))
+                .map(|msg| {
+                    (matches!(msg, ButtonMsg::Clicked)).then(|| CancelConfirmMsg::Confirmed)
+                }),
         )
     }
+
+    pub fn cancel_confirm_text(
+        left: Option<T>,
+        right: T,
+    ) -> CancelConfirm<
+        T,
+        impl Fn(ButtonMsg) -> Option<CancelConfirmMsg>,
+        impl Fn(ButtonMsg) -> Option<CancelConfirmMsg>,
+    >
+    where
+        T: AsRef<str>,
+    {
+        let (left, right_size_factor) = if let Some(verb) = left {
+            (Button::with_text(verb), 1)
+        } else {
+            (Button::with_icon(theme::ICON_CANCEL), 2)
+        };
+        let right = Button::with_text(right).styled(theme::button_confirm());
+
+        Self::cancel_confirm(left, right, right_size_factor)
+    }
+
+    pub fn cancel_info_confirm(
```

Comment:
please modify this to do the two-row placement (as in Ethereum)

otherwise it can probably stay the same? (wrt event handling and the msg type)

Reply:
Fixed in 3b2cd3a76bab5f9ec905a5914007dbbcec514012.

![emu00000000](https://user-images.githubusercontent.com/85857/174164074-ebcdd1e8-529e-4623-8ddc-f7331a875ea6.png)